mlc_llm/quantization/smoothquant_utils.py

```diff
@@ -0,0 +1,322 @@
+
+
+def to_device(params, device):
+    return [param.copyto(device) for param in params]
+
+
+def _accumulate_outlier_stat(stat, data):
+    if stat is None:
+        stat = [[] for i in range(len(data))]
+    for idx, out in enumerate(data):
+        stat[idx].append(out.numpy())
+    return stat
+
+
+def _accumulate_act_outlier_stat(stat, data):
+    a_data = data[::2]
+    return _accumulate_outlier_stat(stat, a_data)
+
+
+def _accumulate_weight_outlier_stat(stat, data):
+    # Optimization step: no need to accumulate weights for each new element in dataset since
+    # weights are the same.
+    if stat is not None:
+        return stat
+    w_data = data[1::2]
+    return _accumulate_outlier_stat(stat, w_data)
+
+
+def _calculate_scale_params(
+    func_name: str,
+    stats,
+    config: Dict[str, Any],
+    dev: tvm.runtime.Device,
+):
+    if stats[func_name] is None:
+        return {}
+
+    # scales = act_scales.pow(alpha) / weight_scales.pow(1-alpha)
+    alpha = config["alpha"]
+    a_stat, w_stat = stats[func_name]
+    assert len(a_stat) == len(w_stat)
+    idx = 0
+    scale_params = {}
+    for a_element, w_elemet in zip(a_stat, w_stat):
+        assert a_element.shape == w_elemet.shape
+        assert len(a_element.shape) == 1
+        scales = np.power(a_element, alpha) / np.power(w_elemet, 1 - alpha)
+        if scales.size - np.count_nonzero(scales) > 0:
+            print("Warning: Smoothing: scales have zero value")
+            scales = np.ones_like(scales)
+            assert False, "Not supported case: please, add more elements in dataset. Otherwise, NaNs in output are possible."
+        scale_params[f"sq_scale_{idx}"] = tvm.nd.array(scales, dev)
+        scale_params[f"sq_scale_{idx+1}"] = tvm.nd.array(scales, dev)
+        idx += 2
+
+    return scale_params
+
+
+def _calculate_quant_scale_params(
+        func_name: str,
+        stats,
+        config: Dict[str, Any],
+        dev: tvm.runtime.Device,
+):
+    if stats[func_name] is None:
+        return {}
+
+    a_dtype = config["adtype"]
+    w_dtype = config["wdtype"]
+
+    idx = 0
+    scale_params = {}
+    for a_element, w_element in zip(*stats[func_name]):
+        a_scale = np.max(a_element) / a_element.dtype.type(np.iinfo(a_dtype).max)
+        w_scale = np.max(w_element) / w_element.dtype.type(np.iinfo(w_dtype).max)
+        scale_params[f"sq_scale_{idx}"] = tvm.nd.array(np.array([a_scale]), dev)
+        scale_params[f"sq_scale_{idx+1}"] = tvm.nd.array(np.array([w_scale]), dev)
+        idx += 2
+
+    return scale_params
+
+
+def _smooth(
+    mod: tvm.IRModule,
+    params: List[tvm.nd.NDArray],
+    funcs: List[str],
+    dataset: List[tvm.nd.NDArray],
+    config: Dict[str, Any],
+):
+    mod = mlc_llm.transform.SmoothQuantAnnotator()(mod)
+    stat_mod = mlc_llm.transform.SmoothQuantStatCollector()(mod)
+    stat_mod = mlc_llm.transform.FuseTransposeMatmul()(stat_mod)
+
+    prefill, decode, kvc, _, _ = get_runtime_func(funcs, stat_mod)
```

Comment:
Very good question! I've been thinking about it too...

At first glance each task (LLM, CV-task etc) has its own pipeline.
For example: pipeline for LLM includes run of encoder + several runs of decoder. Pipeline for Computer Vision task (resnet, inception etc) includes single run of "main" function (as an example). Unit tests can have different set of Relax functions.

I thought about to provide "simple" API for the user. Thus user can configure its own pipeline and run it. But now I don't have such solution. So, for me this is still open question how to make **generic** pipeline.

As for your example with matmul test only: earlier in this place I had a code that checked presence of "main" functions in IRModule. In case of presence I called another one pipeline (for unit tests). But anyway this is ugly approach and not generic..

So, the absence of such **"generic"** pipeline is the main reason why I decided to make this PR to mlc-llm, not to TVM/unity